File: Algorithms/Easy/303.Range-Sum-Query-Immutable/RangeSumQueryImmutable.py
# Space complexity: O(n)
class NumArray:
    # Time complexity: O(n)
    def __init__(self, nums: list[int]):
        self.prefix = [0] * (len(nums) + 1)
        for i in range(len(nums)):
            self.prefix[i + 1] = self.prefix[i] + nums[i]

    # Time complexity: O(1)
    def sumRange(self, left: int, right: int) -> int:
        return self.prefix[right + 1] - self.prefix[left]

File: Algorithms/Easy/303.Range-Sum-Query-Immutable/RangeSumQueryImmutable.go
package easy

type NumArray struct {
	arr []int
}

// Time complexity: O(n)
func Constructor(nums []int) NumArray {
	res := NumArray{
		arr: make([]int, len(nums)+1),
	}

	for i := range nums {
		res.arr[i+1] = res.arr[i] + nums[i]
	}

	return res
}

// Time complexity: O(1)
func (this *NumArray) SumRange(left int, right int) int {
	return this.arr[right+1] - this.arr[left]
}
